apps/sms/services/js/service_manager.js

```diff
@@ -0,0 +1,239 @@
+  };
+
+  const UPGRADE_KEY = 'upgrade-key';
+
+  // Allow old clients to gracefully finish all pending operations before
+  // we destroy them.
+  var ConversationServiceInitializer = {
+    initialize(upgradeKey) {
+      if (upgradeKey) {
+        ConversationClient.init(
+          upgradeKey,
+          new SharedWorker('services/js/conversation/conversation_service.js')
+        );
+      } else {
+        ConversationService.init();
+        ConversationClient.init(App.instanceId, exports);
+      }
+    },
+
+    upgrade(upgradeKey) {
+      var destroyPromise = ConversationClient.destroy().then(() => {
+        ConversationService.destroyService();
+        delete exports.ConversationService;
+
+        debug('Conversation service and client are unloaded');
+      });
+
+      this.initialize(upgradeKey);
+
+      return destroyPromise;
+    }
+  };
+
+  var MessagingServiceInitializer = {
+    initialize(upgradeKey) {
+      if (upgradeKey) {
+        MessagingClient.init(
+          upgradeKey,
+          new SharedWorker('/services/js/messaging/messaging_service.js')
+        );
+      } else {
+        MessagingService.init();
+        MessagingClient.init(App.instanceId, exports);
+      }
+    },
+
+    upgrade(upgradeKey) {
+      var destroyPromise = MessagingClient.destroy().then(() => {
+        MessagingService.destroyService();
+        delete exports.MessagingService;
+
+        debug('Messaging service and client are unloaded');
+      });
+
+      this.initialize(upgradeKey);
+
+      return destroyPromise;
+    }
+  };
+
+  var MozMobileConnectionsServiceInitializer = {
+    initialize(upgradeKey) {
+      if (upgradeKey) {
+        getShimEndpoint(upgradeKey).then((endpoint) => {
+          MozMobileConnectionsClient.init(upgradeKey, endpoint);
+        });
+      } else {
+        MozMobileConnectionsClient.init(App.instanceId, exports);
+      }
+    },
+
+    upgrade(upgradeKey) {
+      var destroyPromise = MozMobileConnectionsClient.destroy().then(() => {
+        MozMobileConnectionsShim.destroyService();
+        delete exports.MozMobileConnectionsShim;
+
+        debug('MozMobileConnections shim and client are unloaded');
+      });
+
+      this.initialize(upgradeKey);
+
+      return destroyPromise;
+    }
+  };
+
+  var MozSettingsServiceInitializer = {
+    initialize() {},
+    upgrade() {
+      return new Promise((resolve) => {
+        MozSettingsShim.destroyService();
+        delete exports.MozSettingsShim;
+        resolve();
+
+        debug('MozSettings shim is unloaded');
+      });
+    }
+  };
+
+  const serviceInitializers = new Map([
+    ['conversation', ConversationServiceInitializer],
+    ['messaging', MessagingServiceInitializer],
+    ['moz-mobile-connections', MozMobileConnectionsServiceInitializer],
+    ['moz-settings', MozSettingsServiceInitializer]
+  ]);
+
+  function initializeShims(endpoint, upgradeKey) {
+    if (endpoint.MozMobileMessageShim) {
```

Comment:
Yep, why not, we can have an array with all related info (ServiceClass and related MozAPI)